### PR TITLE
Deprecate symeig

### PR DESCRIPTION
**Pull request by lezcano**
This one had a tricky usage of `torch.symeig` that had to be replaced. I tested the replacement locally though.

### CODE REVIEW

File: torch/distributions/constraints.py [hunk 1]
```diff
@@ -463,13 +463,11 @@ class _PositiveDefinite(Constraint):
...
+        # Assumes that matrix or matrices in value are symmetric.
+        # There are also non-symmetric positive definite matrices.
         batch_shape = value.unsqueeze(0).shape[:-2]
-        # TODO: replace with batched linear algebra routine when one becomes available
-        # note that `symeig()` returns eigenvalues in ascending order
-        flattened_value = value.reshape((-1,) + matrix_shape)
-        return torch.stack([v.symeig(eigenvectors=False)[0][:1] > 0.0
-                            for v in flattened_value]).view(batch_shape)
+        positive = torch.linalg.eigvalsh(value, UPLO="U")[..., :1] > 0.0
+        return positive.view(batch_shape)
```
Comment (Lezcano) [inline torch/distributions/constraints.py:470]: @IvanYashchuk could you double-check this one?

Comment (IvanYashchuk) [reply torch/distributions/constraints.py:470]: I think here we can use `torch.linalg.cholesky_ex`.
```python
_, info = torch.linalg.cholesky_ex(value)
return info == 0
```
If `info` is larger than 0 then the input (symmetric) matrix is not positive definite numerically.

File: torch/distributions/constraints.py [hunk 2]
```diff
@@ -463,13 +463,11 @@ class _PositiveDefinite(Constraint):
     event_dim = 2
 
     def check(self, value):
-        matrix_shape = value.shape[-2:]
+        # Assumes that matrix or matrices in value are symmetric.
+        # There are also non-symmetric positive definite matrices.
```
Comment (IvanYashchuk) [inline torch/distributions/constraints.py:467]: From Wikipedia:
>  a symmetric n x n real matrix M is said to be positive-definite ...

and then
> Some authors use more general definitions of definiteness, including some non-symmetric real matrices, or non-Hermitian complex ones.

I think for non-symmetric that would be a "generalized positive-definite" matrix 😀

File: torch/distributions/constraints.py (hunk 1 above)
Comment (Lezcano) [reply torch/distributions/constraints.py]: I added it but left a small comment regarding the fact that we assume that the input matrix is symmetric, as before it just used the upper triangular part of the matrix and now we use the lower triangular part (which, btw, is not documented, (although perhaps it's better not to document it))

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -2213,11 +2213,39 @@ std::tuple<Tensor, Tensor> _symeig_helper_cpu(const Tensor& self, bool eigenvect
 }
 
 std::tuple<Tensor, Tensor> symeig(const Tensor& self, bool eigenvectors, bool upper) {
+  TORCH_WARN_ONCE(
+    "torch.symeig is deprecated in favor of torch.linalg.eigh and will be removed in a future ",
+    "PyTorch release.\n",
+    "The default behavior has changed grom using the upper triangular portion of the matrix by default ",
+    "to using the lower triangular portion. This should only matter if the matrix is not symmetric, ",
+    "which is not recommended.\n",
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2221]: why is in not recommended? seems like this should just be removed?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 4]
```diff
@@ -2213,11 +2213,39 @@ std::tuple<Tensor, Tensor> _symeig_helper_cpu(const Tensor& self, bool eigenvect
...
+  );
   squareCheckInputs(self);
   return at::_symeig_helper(self, eigenvectors, upper);
 }
 
 std::tuple<Tensor&, Tensor&> symeig_out(const Tensor& self, bool eigenvectors, bool upper, Tensor& vals, Tensor& vecs) {
+  TORCH_WARN_ONCE(
+    "torch.symeig is deprecated in favor of torch.linalg.eigh and will be removed in a future ",
+    "PyTorch release.\n",
+    "The default behavior has changed grom using the upper triangular portion of the matrix by default ",
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2238]: Same "grom" here

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -2213,11 +2213,39 @@ std::tuple<Tensor, Tensor> _symeig_helper_cpu(const Tensor& self, bool eigenvect
...
+    "to using the lower triangular portion. This should only matter if the matrix is not symmetric, ",
+    "which is not recommended.\n",
+    "L, _ = torch.symeig(A, upper=upper)\n",
```
Comment (mruberry) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:2241]: Really nice upgrade guide

File: torch/_linalg_utils.py [hunk 6]
```diff
@@ -88,14 +88,12 @@ def basis(A):
     return Q
 
 
-def symeig(A: Tensor, largest: Optional[bool] = False, eigenvectors: Optional[bool] = True) -> Tuple[Tensor, Tensor]:
+def symeig(A: Tensor, largest: Optional[bool] = False) -> Tuple[Tensor, Tensor]:
```
Comment (mruberry) [inline torch/_linalg_utils.py:91]: Is this function actually used anywhere?

Comment (mruberry) [reply torch/_linalg_utils.py:91]: This would change its signature, right?

File: torch/_lobpcg.py [hunk 7]
```diff
@@ -997,7 +997,7 @@ def _get_svqb(self,
         # The original algorithm 4 from [DuerschPhD2015].
         d_col = (d ** -0.5).reshape(d.shape[0], 1)
         DUBUD = (UBU * d_col) * _utils.transpose(d_col)
-        E, Z = _utils.symeig(DUBUD, eigenvectors=True)
+        E, Z = _utils.symeig(DUBUD)
```
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_lobpcg.py:1000]: Checking internally, it looks like the only uses of this function are in this file. Probably a good chance to kill it?

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -8625,6 +8625,28 @@ def merge_dicts(*dicts):
 
 If :attr:`upper` is ``False``, then lower triangular portion is used.
 
+.. warning::
+
+    :func:`torch.symeig` is deprecated in favor of :func:`torch.linalg.eigh`
+    and will be removed in a future PyTorch release. The default behavior has changed
+    from using the upper triangular portion of the matrix by default to using the
+    lower triangular portion. This should only matter if this function is used with
+    non-symmetric matrices, which is not recommended.
```
Comment (mruberry) [inline torch/_torch_docs.py:8634]: "not recommended" parenethetical here should probably be killed

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 9]
```diff
@@ -2346,8 +2346,8 @@ Tensor symeig_backward(const std::vector<torch::autograd::Variable> &grads, cons
   // leads to stable gradient updates, and retains symmetry of the updated matrix if it
   // were updated by a gradient based algorithm.
   TORCH_CHECK(eigenvectors,
-           "symeig_backward: Setting eigenvectors to false in torch.symeig doesn't compute eigenvectors ",
-           "and hence we cannot compute backward. Please use torch.symeig(eigenvectors=True)");
+           "symeig_backward: torch.symeig(A, eigenvectors=False) is not differentiable. ",
+           "Please use torch.linalg.eigvalsh(A)");
```
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:2350]: "Use torch.linalg.eigvalsh(A), instead"

File: torch/distributions/constraints.py [hunk 10]
```diff
@@ -463,13 +463,9 @@ class _PositiveDefinite(Constraint):
...
-        matrix_shape = value.shape[-2:]
-        batch_shape = value.unsqueeze(0).shape[:-2]
-        # TODO: replace with batched linear algebra routine when one becomes available
-        # note that `symeig()` returns eigenvalues in ascending order
-        flattened_value = value.reshape((-1,) + matrix_shape)
-        return torch.stack([v.symeig(eigenvectors=False)[0][:1] > 0.0
-                            for v in flattened_value]).view(batch_shape)
+        # Assumes that the matrix or batch of matrices in value are symmetric
+        # info == 0 means no error, that is, it's SPD
+        return torch.linalg.cholesky_ex(value).info.eq(0).unsqueeze(0)
```
Comment (mruberry) [inline torch/distributions/constraints.py:468]: Would you file an issue that we should consider adding fast, explicit checks for matrix properties, like whether they're positive definitive, @Lezcano?

cc @Balandat

Comment (mruberry) [reply torch/distributions/constraints.py:468]: @IvanYashchuk would you sanity check this, too?

**Review by mruberry — COMMENTED** [review]
This looks good overall, @Lezcano; I made a few inline comments with suggestions and questions

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:2221]: I certainly think so but numpy-compat...

This function just uses the upper or lower half of the matrix, because it's assumed to be a symmetric matrix, but then the gradien that we return is not upper or lower triangular, but symmetric.This is inconsistent, as a clever user could want to represent two different symmetric matrices with zeros in the diagonal in the same matrix. I this user then used this construction in a differentiable environment, the gradients would be wrong, even though what they were doing seemed semantically right...

In my opinion, we should ask for symmetric matrices, and then user the upper or lower triangular part of the matrix internally, or symmetrise it first to remove possible errors, but imo, these things should not be exposed to the user.

That being said, numpy-compat...

File: torch/_lobpcg.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_lobpcg.py:1000]: This function gets the eigenvalues and reorders them from largest to smallest or the other way around. I think it's a reasonable helper function to have around.

We could further ask the question of whether this is something that we would like to implement internally in `eigh` (I think so), but that's a question for another time.

File: torch/_torch_docs.py (hunk 8 above)
Comment (Lezcano) [reply torch/_torch_docs.py:8634]: I'll delete the whole sentence since, as described in https://github.com/pytorch/pytorch/pull/57732#discussion_r629202509 we do not really support this behaviour.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 9 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2350]: Living the last few years in the UK has gotten the best of me.

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2350]: There are a number of constructions similar to this one in other errors. As it's not critical, I will not touch the others, as doing so would conflict with other PRs that have already been imported to phabricator.

**Comment by Lezcano** [comment]
@mruberry I've answered the questions and incorporated the suggestions.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (mruberry) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:2354]: I appreciate there's some nuance here and we can dig into it more offline. I just don't like the vagueness of things being not "recommended." Like, either we support things or we don't, and if we don't support them the behavior should be undefined, or defined but subject to change and curse the user who tries to rely on it. 

That said, this is far from the worst thing in our docs today; we can improve it later

File: torch/_lobpcg.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_lobpcg.py:1000]: OK

**Review by mruberry — APPROVED** [review]
Cool! I left a couple notes but let's not do any more changes in this PR

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:2354]: I forgot to remove these from the comments, I'll do this now.
I also filled an issue describing this problem: https://github.com/pytorch/pytorch/issues/58037

**Comment by Lezcano** [comment]
@mruberry This should be ready
